plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -68,46 +108,106 @@ class bp_connection_manager {
+         }
+      }
+   }
+
+   // Called when configured bp peer key changes
+   void update_bp_producer_peers(const chain::controller& cc, const std::string& server_address) {
+      assert(!config.my_bp_peer_accounts.empty());
+      fc::lock_guard gm(mtx);
+      fc::lock_guard g(gossip_bps.mtx);
+      // normally only one bp peer account except in testing scenarios or test chains
+      for (const auto& e : config.my_bp_peer_accounts) { // my_bp_peer_accounts not modified after plugin startup
+         std::optional<peer_info_t> peer_info = cc.get_peer_info(e);
+         if (peer_info && peer_info->key) {
+            // update initial so always an active one
+            gossip_bp_peers_message::bp_peer signed_empty{.producer_name = e}; // .server_address not set for initial message
+            signed_empty.sig = self()->sign_compact(*peer_info->key, signed_empty.digest());
+            EOS_ASSERT(signed_empty.sig != signature_type{}, chain::plugin_config_exception,
+                       "Unable to sign empty gossip bp peer, private key not found for ${k}", ("k", peer_info->key->to_string({})));
+            initial_gossip_msg_factory.set_initial_send_buffer(signed_empty);
+            // update gossip_bps
+            auto& prod_idx = gossip_bps.index.get<by_producer>();
+            gossip_bp_peers_message::bp_peer peer{.producer_name = e, .server_address = server_address};
+            peer.sig = self()->sign_compact(*peer_info->key, peer.digest());
+            EOS_ASSERT(peer.sig != signature_type{}, chain::plugin_config_exception,
+                       "Unable to sign bp peer ${p}, private key not found for ${k}", ("p", peer.producer_name)("k", peer_info->key->to_string({})));
+            if (auto i = prod_idx.find(boost::make_tuple(e, boost::cref(server_address))); i != prod_idx.end()) {
+               gossip_bps.index.modify(i, [&peer](auto& v) {
+                  v.sig = peer.sig;
+               });
+            } else {
+               gossip_bps.index.emplace(peer);
+            }
+         }
+      }
    }
 
    // Only called from connection strand and the connection constructor
-   void mark_bp_connection(Connection* conn) const {
-      /// mark an connection as a bp connection if it connects to an address in the bp peer list, so that the connection
-      /// won't be subject to the limit of max_client_count.
+   void mark_configured_bp_connection(Connection* conn) const {
+      /// mark an connection as a configured bp connection if it connects to an address in the bp peer list,
+      /// so that the connection won't be subject to the limit of max_client_count.
       auto space_pos = conn->log_p2p_address.find(' ');
       // log_p2p_address always has a trailing hex like `localhost:9877 - bc3f55b`
       std::string addr = conn->log_p2p_address.substr(0, space_pos);
       if (config.bp_peer_accounts.count(addr)) {
```

Comment:
I'm not sure howw reliable that is, wouldn't this fail if `set_configured_bp_peers` was called with strings having a space after the coma, or at the end of the string after the ip address.
Why not store a std::pair of `host, port` as returned from `net_utils::split_host_port_type()` instead of a string?

Reply:
https://github.com/AntelopeIO/spring/pull/1265/commits/3b8145f94d14b9ea06e3b7696ed1b84e910b7ffe